### PR TITLE
Memory management cleanup

### PR DESCRIPTION
**Pull request by roystgnr**
Refs #7714

## Reason
There's still quite a lot of manual memory allocation/deallocation in MOOSE, a lot of it unnecessary.  This PR doesn't actually cover the allocation issues that sent me down this rabbit hole, but it should fix up most of the raw pointers and raw allocation that are easy to change.

## Design
* Replace raw (manually allocated/deallocated) pointers with `unique_ptr` where appropriate
* Replace raw `new` with `make_unique` or `make_shared` as appropriate
* Use the "cast reference via a pointer" trick that we use for libMesh mesh iterators, to get rid of the need for (and then get rid of) identical-but-with-more-const copies of pointers in `Assembly`

## Impact
I've tried to limit changes to places where encapsulation allowed me to avoid any public API changes.  There are changes to *protected* member types, though, so it's not impossible that subclasses using those might fail to compile.

### CODE REVIEW

File: framework/src/transfers/MultiAppGeneralFieldShapeEvaluationTransfer.C [hunk 1]
```diff
@@ -72,11 +72,11 @@ MultiAppGeneralFieldShapeEvaluationTransfer::buildMeshFunctions(
     System & from_sys = from_var.sys().system();
     unsigned int from_var_num = from_sys.variable_number(getFromVarName(var_index));
 
-    std::shared_ptr<MeshFunction> from_func;
-    from_func.reset(new MeshFunction(getEquationSystem(from_problem, _displaced_source_mesh),
-                                     *from_sys.current_local_solution,
-                                     from_sys.get_dof_map(),
-                                     from_var_num));
+    std::shared_ptr<MeshFunction> from_func =
+        std::make_shared<MeshFunction>(getEquationSystem(from_problem, _displaced_source_mesh),
```
Comment (GiudGiud) [inline framework/src/transfers/MultiAppGeneralFieldShapeEvaluationTransfer.C:76]: same

File: framework/src/transfers/MultiAppShapeEvaluationTransfer.C [hunk 2]
```diff
@@ -245,11 +245,11 @@ MultiAppShapeEvaluationTransfer::transferVariable(unsigned int i)
     System & from_sys = from_var.sys().system();
     unsigned int from_var_num = from_sys.variable_number(from_var.name());
 
-    std::shared_ptr<MeshFunction> from_func;
-    from_func.reset(new MeshFunction(getEquationSystem(from_problem, _displaced_source_mesh),
-                                     *from_sys.current_local_solution,
-                                     from_sys.get_dof_map(),
-                                     from_var_num));
+    std::shared_ptr<MeshFunction> from_func =
+        std::make_shared<MeshFunction>(getEquationSystem(from_problem, _displaced_source_mesh),
```
[the remark this answers is not in the document]
Comment (roystgnr) [reply framework/src/transfers/MultiAppShapeEvaluationTransfer.C:249]: Trying to make minimal changes, I guess, but you're right; there's no need for a smart pointer here at all.

**Comment by roystgnr** [comment]
Bah.  Is 14.0.0 not a safe version of clang-format for me to use?  I thought I was being clever by finally just auto-running that when saving files.

**Review by GiudGiud — COMMENTED** [review]
added the valgrind recipes

**Review by GiudGiud — APPROVED** [review]
nice cleanup. This will surely help us do better at memory management by showing the better practices

**Pull request merged**